### PR TITLE
Update Prometheus to latest version

### PR DESCRIPTION
**Pull request by replay**
This PR updates the Prometheus version vendored in Cortex to the latest version.

With the new Prometheus version the `LabelValues()` call on the `Querier` interface now has an additional `matchers` parameter (introduced [here](https://github.com/prometheus/prometheus/pull/8400)). 
Unfortunately we won't be able to directly take advantage of this new parameter here, because in order to use it this struct which gets imported from Thanos will also need to be updated: https://github.com/cortexproject/cortex/blob/d385f47f011ddbf436f5695928e18492834b5d92/vendor/github.com/thanos-io/thanos/pkg/store/storepb/rpc.pb.go#L503

We also can't update the Prometheus version in Thanos first, because Thanos imports Querier implementations from Cortex, so if we'd update Prometheus in Thanos before updating it in Cortex then the imported Querier implementations wouldn't match the interface defined in Prometheus anymore. 
My plan to solve this is to first update the Prometheus version in Cortex, without taking advantage of that new parameter in the `LabelValues()` call. Then update the Prometheus and the Cortex versions in Thanos. Then make another PR to Cortex which takes advantage of that new parameter.

The main thing I'm uncertain about here is how to deal with the fact that `v1.NewAPI` ([here](https://github.com/cortexproject/cortex/blob/bf46250493da0d5c2ed988c964516f56c7a93a24/pkg/api/handlers.go#L190)) now requires the second argument to be a `storage.Storage`, previously it was a `storage.SampleAndChunkQueryable`. Since the last argument that we pass into `v1.NewAPI()` ([here](https://github.com/cortexproject/cortex/blob/bf46250493da0d5c2ed988c964516f56c7a93a24/pkg/api/handlers.go#L210)) is now hard-coded to `false` I think it is acceptable to simply add the required methods to `errorTranslateQueryable` without implementing them, knowing that they will never be called, but I'd definitely be open to suggestions about how to do that in a better way.

### CODE REVIEW

**Comment by pstibrany** [comment]
> The main thing I'm uncertain about here is how to deal with the fact that `v1.NewAPI` ([here](https://github.com/cortexproject/cortex/blob/bf46250493da0d5c2ed988c964516f56c7a93a24/pkg/api/handlers.go#L190)) now requires the second argument to be a `storage.Storage`, previously it was a `storage.SampleAndChunkQueryable`. Since the last argument that we pass into `v1.NewAPI()` ([here](https://github.com/cortexproject/cortex/blob/bf46250493da0d5c2ed988c964516f56c7a93a24/pkg/api/handlers.go#L210)) is now hard-coded to `false` I think it is acceptable to simply add the required methods to `errorTranslateQueryable` without implementing them, knowing that they will never be called, but I'd definitely be open to suggestions about how to do that in a better way.

I would suggest to keep `errorTranslateQueryable` as is (ie. reverting your change), and introduce new `readonlyStorage` type that embeds `storage.SampleAndChunkQueryable`, and adds additional missing methods from Storage interface. Those missing methods should also return errors where appropriate (esp. Appender part). WDYT?

**Comment by replay** [comment]
> I would suggest to keep errorTranslateQueryable as is (ie. reverting your change), and introduce new readOnlyStorage type that embeds storage.SampleAndChunkQueryable, and adds additional missing methods from Storage interface. Those missing methods should also return errors where appropriate (esp. Appender part). WDYT?

That sounds better indeed, I'll do that, thx

**Comment by replay** [comment]
@pstibrany i updated it, thanks for the suggestion

**Comment by codesome** [comment]
This might need to go into the next release to include [this fix](https://github.com/prometheus/prometheus/commit/b18fde996e29062b62bc033e8db3b3edc4887d04) (affects `timestamp()` function with @ modifier).

**Review by pstibrany — APPROVED** [review]
It is unfortunate that NewAPI takes `Storage`, which requires us to implement extra methods.  I've sent PR https://github.com/prometheus/prometheus/pull/8488 to Prometheus to be able to pass queryable and appendable separately. If it is merged quickly, then I'm fine with taking shortcuts in *this* PR and just implement bare-minimum (as it is now). However, if https://github.com/prometheus/prometheus/pull/8488 isn't merged, then we should implement "errAppender" to return errors, as current implementation of `readOnlyStorage.Appender()` returning `nil` may cause panics. It's also not clear what we should be returning in `StartTime`, although `API` doesn't use this method currently.

Apart from that, this PR looks good.

**Comment by roidelapluie** [comment]
I will merge it on the other side :)

**Comment by pstibrany** [comment]
Since Prometheus PR 8488 is now merged, let's update Prometheus in this PR.

**Comment by replay** [comment]
Thx for that Prometheus change @pstibrany . I have update Prometheus in this branch again, and I also rebased onto the latest master.

File: pkg/api/handlers.go [hunk 1]
```diff
@@ -190,6 +190,7 @@ func NewQuerierHandler(
 	api := v1.NewAPI(
 		engine,
 		errorTranslateQueryable{queryable}, // Translate errors to errors expected by API.
+		nil,
```
Comment (pracucci) [inline pkg/api/handlers.go:193]: Let's just explain why:

```suggestion
		nil, // No remote write support.
```

File: pkg/querier/blocks_store_queryable.go [hunk 2]
```diff
@@ -870,7 +871,8 @@ func createLabelNamesRequest(minT, maxT int64, blockIDs []ulid.ULID) (*storepb.L
 	return req, nil
 }
 
-func createLabelValuesRequest(minT, maxT int64, label string, blockIDs []ulid.ULID) (*storepb.LabelValuesRequest, error) {
+func createLabelValuesRequest(minT, maxT int64, label string, blockIDs []ulid.ULID, matchers ...*labels.Matcher) (*storepb.LabelValuesRequest, error) {
+	// TODO(replay): add matchers to LabelValuesRequest once it has that property
```
Comment (pracucci) [inline pkg/querier/blocks_store_queryable.go:875]: LGTM, considering the store-gateway doesn't support it yet.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Review by pstibrany — APPROVED** [review]
Good job, thank you!

**Pull request merged**